include/RAJA/pattern/WorkGroup/WorkRunner.hpp

```diff
@@ -0,0 +1,295 @@
+  template < typename segment_type, typename loop_type >
+  using holder_type = HoldForall<forall_exec_policy, segment_type, loop_type,
+                                 index_type, Args...>;
+
+  // The policy indicating where the call function is invoked
+  // in this case the values are called on the host in a loop
+  using vtable_exec_policy = RAJA::loop_work;
+
+  // runner interfaces with storage to enqueue so the runner can get
+  // information from the segment and loop at enqueue time
+  template < typename WorkContainer, typename segment_T, typename loop_T >
+  inline void enqueue(WorkContainer& storage, segment_T&& seg, loop_T&& loop)
+  {
+    using holder = holder_type<camp::decay<segment_T>, camp::decay<loop_T>>;
+
+    storage.template emplace<holder>(
+        get_Vtable<holder, vtable_type>(vtable_exec_policy{}),
+        std::forward<segment_T>(seg), std::forward<loop_T>(loop));
+  }
+
+  // clear any state so ready to be destroyed or reused
+  void clear()
+  { }
+
+  // no extra storage required here
+  using per_run_storage = int;
+};
+
+/*!
+ * Runs work in a storage container in order using forall
+ */
+template <typename FORALL_EXEC_POLICY,
+          typename EXEC_POLICY_T,
+          typename ORDER_POLICY_T,
+          typename ALLOCATOR_T,
+          typename INDEX_T,
+          typename ... Args>
+struct WorkRunnerForallOrdered
+    : WorkRunnerForallOrdered_base<
+      FORALL_EXEC_POLICY,
+      EXEC_POLICY_T,
+      ORDER_POLICY_T,
+      ALLOCATOR_T,
+      INDEX_T,
+      Args...>
+{
+  using base = WorkRunnerForallOrdered_base<
+      FORALL_EXEC_POLICY,
+      EXEC_POLICY_T,
+      ORDER_POLICY_T,
+      ALLOCATOR_T,
+      INDEX_T,
+      Args...>;
+  using base::base;
+
+  // run the loops using forall in the order that they were enqueued
+  template < typename WorkContainer >
+  typename base::per_run_storage run(WorkContainer const& storage, Args... args) const
+  {
+    using value_type = typename WorkContainer::value_type;
+
+    typename base::per_run_storage run_storage{};
+
+    auto end = storage.end();
+    for (auto iter = storage.begin(); iter != end; ++iter) {
+      value_type::call(&*iter, args...);
+    }
+
+    return run_storage;
+  }
+};
+
+/*!
+ * Runs work in a storage container in reverse order using forall
+ */
+template <typename FORALL_EXEC_POLICY,
+          typename EXEC_POLICY_T,
+          typename ORDER_POLICY_T,
+          typename ALLOCATOR_T,
+          typename INDEX_T,
+          typename ... Args>
+struct WorkRunnerForallReverse
+    : WorkRunnerForallOrdered_base<
+      FORALL_EXEC_POLICY,
+      EXEC_POLICY_T,
+      ORDER_POLICY_T,
+      ALLOCATOR_T,
+      INDEX_T,
+      Args...>
+{
+  using base = WorkRunnerForallOrdered_base<
+      FORALL_EXEC_POLICY,
+      EXEC_POLICY_T,
+      ORDER_POLICY_T,
+      ALLOCATOR_T,
+      INDEX_T,
+      Args...>;
+  using base::base;
+
+  // run the loops using forall in the reverse order to the order they were enqueued
```

Comment:
Why reverse order?

Reply:
This is the generic reverse ordered runner implementation, there is a generic inorder ordered runner above.